lib/yggdrasil/player.ex

```diff
@@ -12,9 +13,16 @@ defmodule Yggdrasil.Player do
     Supervisor.add_player user_id, [owner_pid, push_msg]
   end
 
-  def run_cmd(user_id, cmd = %Message{type: :command}) do
-    player_pid = Registry.get_player user_id
-    GenServer.cast(player_pid, {:run_cmd, cmd})
+  def run_cmd(user_id, cmd) do
+    # horrible terrible temporary implementation
+    # room would eventually do all of this
+    ctxt = %Yggdrasil.Room.Context{ player: user_id }
+    ctxt = Command.execute cmd, ctxt
+
+    Enum.each ctxt.actions,
+      fn {:notify, player, message} ->
+        notify(player, message)
+      end
```

Comment:
Yeah, actually very similar to what I was proposing just instead of executing the command then looping through the context the context would be built first then sent to the room and the room applies it by looping and performs the mutation. 

```elixir
# probably should be build
def build(cmd, ctx) do
  ctx
  |> move_player(cmd.exit)
  |> notify_player("You head #{cmd.exit}")
end

# player module
# by starting in the process instead of the caller
# we know the player's state and the room it's in I think it makes sense to put it here
def handle_cast({:run_cmd, cmd}, _, state) do
  ctx = Game.Context.new
  |> add_character(state.character)
  |> add_room(state.room)

  # nothing has been done at this point, no side effects.
  # just a pure data structure
  # this step just adds stuff to the context 
  ctx = Command.build(cmd, ctx)

  Room.run ctx
end

# room module

# absolutely agree with you on the need for halting the action chain,
# quick example
def handle_cast({:apply_context, ctx}, _, state} do
  Enum.each ctx.actions, fn action ->
    perform_action(action, state) 
  end
  ...
end

def run(ctx) do
  GenServer.cast(ctx.room, {:apply_context, ctx})
  ...
end

# i realize this take a user_id but not sure we will want that from a room's stand point
def action({:notify, pid, msg}, state) do
  Player.notify pid, msg
  :ok
end

def action({:move, exit}, state) do
  case Enum.any? state.exits, fn ex -> ex == exit end do 
    true -> :ok
    false -> {:error, "#{exit} doesn't exist in this room"}
  end
end
```
I don't know like I'm good with what you are doing as well, mainly was trying to eliminate the commands in the rooms, it seems like they could operate entirely off of context which means we could test the commands in isolation as they just produce a context, and we could test rooms without a command and have fine control over the context during testing to exercise all the actions.

So for now I would just stay on the path you are own for now, as we get there will be more for us to point at and articulate different designs. I think the rooms are the part where this all will shake out.